src/ol/source/Vector.js

```diff
@@ -602,7 +602,16 @@ class VectorSource extends Source {
    */
   forEachFeature(callback) {
     if (this.featuresRtree_) {
-      return this.featuresRtree_.forEach(callback);
+      let result = this.featuresRtree_.forEach(callback);
+      if (!result) {
```

Comment:
Shouldn't the `nullGeometryFeatures` always be added to the result, not just when there are no other features?

Reply:
It stops when any callback call returns a  truthy values.
https://github.com/openlayers/openlayers/blob/7c9c2dacf71be831dc3ccd3bae07b003532710d1/src/ol/source/Vector.js#L613-L616

Reply:
But also:

https://github.com/openlayers/openlayers/blob/19bb634a650003bbf067c988a1200312611eba8b/src/ol/source/Vector.js#L595

Reply:
This is only true if a spatial index is used, if it iterates over the `featuresCollection_`, null-geometries are not ignored.

Reply:
I may be missing something, but to me it seems your change changes the behavior of the case where a `featuresRtree_` is available, which is the documented case where null geometries are ignored. Did you mean to change that behavior? If so, you should add a note about it in the upgrade notes, and we need to bump the major version on the next release. If I'm missing something, maybe you can add a test that shows what's currently wrong. Thanks!

Reply:
Yeah, this is exactly it. To me it was rather unexpected behavior to skip some features.

It is documented, but I hadn't read it...